Added link to VSO Health Report in Fido VSOClient output

ahmvdev opened the pull request:
This pull request introduces an update to the output of the Fido results when using the VSOClient. The change adds a link to the latest VSO Health Report, which provides up-to-date status on VSO data retrieval. 

### Changes
The output of the Fido results now includes the followinf string: 
`Latest VSO data retrieval status available at https://docs.virtualsolar.org/wiki/VSOHealthReport`
This will provide users with a direct link to the VSO Health Report for more status on the data  retrieval

### Example Output:
```
Results from 1 Provider:

1 Results from the VSOClient:
Source: http://vso.stanford.edu/cgi-bin/search
Total estimated size: 67.789 Mbyte

       Start Time               End Time        Source Instrument   Wavelength   Provider  Physobs  Wavetype Extent Width Extent Length Extent Type   Size  
                                                                     Angstrom                                                                        Mibyte 
----------------------- ----------------------- ------ ---------- -------------- -------- --------- -------- ------------ ------------- ----------- --------
2012-03-04 00:00:00.000 2012-03-04 00:00:01.000    SDO        AIA 171.0 .. 171.0     JSOC intensity   NARROW         4096          4096    FULLDISK 64.64844

Latest VSO data retrieval status available at https://docs.virtualsolar.org/wiki/VSOHealthReport
```
Feel free to leave any feedback or suggestions on this change. Thanks

### Related Issue:
- Resolves #7853

nabobalis commented on sunpy/net/fido_factory.py:
```diff
@@ -196,6 +196,9 @@ def __str__(self):
             ret += '\n\n'
 
+        ret += "Latest VSO data retrieval status available at https://docs.virtualsolar.org/wiki/VSOHealthReport\n"
+
+
```
```suggestion
```

nabobalis commented:
Thanks for the PR, can you add a changelog entry for this please?

nabobalis commented on sunpy/net/fido_factory.py:
```diff
@@ -196,6 +196,9 @@ def __str__(self):
             ret += '\n'.join(lines[1:])
             ret += '\n\n'
 
+        ret += "Latest VSO data retrieval status available at https://docs.virtualsolar.org/wiki/VSOHealthReport\n"
```
```suggestion
        ret += "VSO Data Provider Health Status available at https://docs.virtualsolar.org/wiki/VSOHealthReport\n"
```

nabobalis commented:
I do wonder if this should be added to the top instead, where we have the "Source: http://vso.stanford.edu/cgi-bin/search" instead?

ahmvdev commented:
Hi, thanks alot for the feedback, I have made the following updates:

- Added changelog entry for the feature
- Updated the string from `Latest VSO data retrieval status available at https://docs.virtualsolar.org/wiki/VSOHealthReport` to `VSO Data Provider Health Status available at https://docs.virtualsolar.org/wiki/VSOHealthReport` and removed the extra line.
- I moved the string one line before `Source: http://vso.stanford.edu/cgi-bin/search`
- Updated the unit test in `test_fido.py` to account for 9 header lines instead of 8. I think that caused the test to fail before.

All tests are passing now.

nabobalis commented on CHANGELOG.rst:
```diff
@@ -55,7 +55,7 @@ Removals
 New Features
 ------------
-
+- Added a link to the VSO Health Report in the Fido Results when using the VSOClient. (`#7853 <https://github.com/sunpy/sunpy/issues/7853>`__)
```
This is unfortunately not how changelog entries are created, please revert this file and follow https://github.com/sunpy/sunpy/blob/main/changelog/README.rst

nabobalis commented:
How does the output look now?

nabobalis commented:
The following online tests are failing:

```
FAILED ../../.tox/py312-online/lib/python3.12/site-packages/sunpy/net/dataretriever/sources/goes.py::sunpy.net.dataretriever.sources.goes.SUVIClient
FAILED ../../.tox/py312-online/lib/python3.12/site-packages/sunpy/net/dataretriever/sources/goes.py::sunpy.net.dataretriever.sources.goes.XRSClient
FAILED ../../.tox/py312-online/lib/python3.12/site-packages/sunpy/net/dataretriever/sources/eve.py::sunpy.net.dataretriever.sources.eve.EVEClient
FAILED ../../.tox/py312-online/lib/python3.12/site-packages/sunpy/net/dataretriever/sources/lyra.py::sunpy.net.dataretriever.sources.lyra.LYRAClient
FAILED ../../.tox/py312-online/lib/python3.12/site-packages/sunpy/net/dataretriever/sources/gong.py::sunpy.net.dataretriever.sources.gong.GONGClient
FAILED ../../.tox/py312-online/lib/python3.12/site-packages/sunpy/net/dataretriever/sources/noaa.py::sunpy.net.dataretriever.sources.noaa.NOAAIndicesClient
FAILED ../../.tox/py312-online/lib/python3.12/site-packages/sunpy/net/dataretriever/sources/fermi_gbm.py::sunpy.net.dataretriever.sources.fermi_gbm.GBMClient
FAILED ../../.tox/py312-online/lib/python3.12/site-packages/sunpy/net/dataretriever/sources/rhessi.py::sunpy.net.dataretriever.sources.rhessi.RHESSIClient
FAILED ../../.tox/py312-online/lib/python3.12/site-packages/sunpy/net/dataretriever/sources/noaa.py::sunpy.net.dataretriever.sources.noaa.NOAAPredictClient
FAILED ../../.tox/py312-online/lib/python3.12/site-packages/sunpy/net/dataretriever/sources/norh.py::sunpy.net.dataretriever.sources.norh.NoRHClient
FAILED ../../docs/whatsnew/5.0.rst::5.0.rst
FAILED ../../docs/how_to/search_multiple_wavelengths.rst::search_multiple_wavelengths.rst
FAILED ../../docs/how_to/search_vso.rst::search_vso.rst
```
They will need to be tracked down to check if they are due to this PR or servers broken upstream.

ahmvdev commented:
> FAILED ../../.tox/py312-online/lib/python3.12/site-packages/sunpy/net/dataretriever/sources/goes.py::sunpy.net.dataretriever.sources.goes.SUVIClient
> FAILED ../../.tox/py312-online/lib/python3.12/site-packages/sunpy/net/dataretriever/sources/goes.py::sunpy.net.dataretriever.sources.goes.XRSClient
> FAILED ../../.tox/py312-online/lib/python3.12/site-packages/sunpy/net/dataretriever/sources/eve.py::sunpy.net.dataretriever.sources.eve.EVEClient
> FAILED ../../.tox/py312-online/lib/python3.12/site-packages/sunpy/net/dataretriever/sources/lyra.py::sunpy.net.dataretriever.sources.lyra.LYRAClient
> FAILED ../../.tox/py312-online/lib/python3.12/site-packages/sunpy/net/dataretriever/sources/gong.py::sunpy.net.dataretriever.sources.gong.GONGClient
> FAILED ../../.tox/py312-online/lib/python3.12/site-packages/sunpy/net/dataretriever/sources/noaa.py::sunpy.net.dataretriever.sources.noaa.NOAAIndicesClient
> FAILED ../../.tox/py312-online/lib/python3.12/site-packages/sunpy/net/dataretriever/sources/fermi_gbm.py::sunpy.net.dataretriever.sources.fermi_gbm.GBMClient
> FAILED ../../.tox/py312-online/lib/python3.12/site-packages/sunpy/net/dataretriever/sources/rhessi.py::sunpy.net.dataretriever.sources.rhessi.RHESSIClient
> FAILED ../../.tox/py312-online/lib/python3.12/site-packages/sunpy/net/dataretriever/sources/noaa.py::sunpy.net.dataretriever.sources.noaa.NOAAPredictClient
> FAILED ../../.tox/py312-online/lib/python3.12/site-packages/sunpy/net/dataretriever/sources/norh.py::sunpy.net.dataretriever.sources.norh.NoRHClient

i figured it out. i think the above tests failed because there was no check in place for `VSOClient` and the Health Report Link was showing for other clients aswell

ahmvdev commented:
> How does the output look now?

Here's an example output:

```Results from 1 Provider:

145 Results from the VSOClient:
VSO Data Provider Health Status available at https://docs.virtualsolar.org/wiki/VSOHealthReport
Source: http://vso.stanford.edu/cgi-bin/search
Total estimated size: 488.147 Gbyte

       Start Time               End Time        Source Instrument   Wavelength   ... Extent Width Extent Length Extent Type    Size    Extra Flags
                                                                     Angstrom    ...                                          Mibyte              
----------------------- ----------------------- ------ ---------- -------------- ... ------------ ------------- ----------- ---------- -----------
2012-03-04 00:00:00.000 2012-03-04 00:09:25.000    SDO        AIA 171.0 .. 171.0 ...         4096          4096    FULLDISK   3103.125           S
2012-03-04 00:09:36.000 2012-03-04 00:19:25.000    SDO        AIA 171.0 .. 171.0 ...         4096          4096    FULLDISK 3232.42188           S
2012-03-04 00:19:36.000 2012-03-04 00:29:25.000    SDO        AIA 171.0 .. 171.0 ...         4096          4096    FULLDISK 3232.42188           S
2012-03-04 00:29:36.000 2012-03-04 00:39:25.000    SDO        AIA 171.0 .. 171.0 ...         4096          4096    FULLDISK 3232.42188           S
2012-03-04 00:39:36.000 2012-03-04 00:49:25.000    SDO        AIA 171.0 .. 171.0 ...         4096          4096    FULLDISK 3232.42188           S
2012-03-04 00:49:36.000 2012-03-04 00:59:25.000    SDO        AIA 171.0 .. 171.0 ...         4096          4096    FULLDISK 3232.42188           S
2012-03-04 00:59:36.000 2012-03-04 01:09:25.000    SDO        AIA 171.0 .. 171.0 ...         4096          4096    FULLDISK 3232.42188           S
                    ...                     ...    ...        ...            ... ...          ...           ...         ...        ...         ...
2012-03-04 22:59:36.000 2012-03-04 23:09:25.000    SDO        AIA 171.0 .. 171.0 ...         4096          4096    FULLDISK 3232.42188           S
2012-03-04 23:09:36.000 2012-03-04 23:19:25.000    SDO        AIA 171.0 .. 171.0 ...         4096          4096    FULLDISK 3232.42188           S
2012-03-04 23:19:36.000 2012-03-04 23:29:25.000    SDO        AIA 171.0 .. 171.0 ...         4096          4096    FULLDISK 3232.42188           S
2012-03-04 23:29:36.000 2012-03-04 23:39:25.000    SDO        AIA 171.0 .. 171.0 ...         4096          4096    FULLDISK 3232.42188           S
2012-03-04 23:39:36.000 2012-03-04 23:49:25.000    SDO        AIA 171.0 .. 171.0 ...         4096          4096    FULLDISK 3232.42188           S
2012-03-04 23:49:36.000 2012-03-04 23:59:25.000    SDO        AIA 171.0 .. 171.0 ...         4096          4096    FULLDISK 3232.42188           S
2012-03-04 23:59:36.000 2012-03-05 00:00:01.000    SDO        AIA 171.0 .. 171.0 ...         4096          4096    FULLDISK  193.94531           S
Length = 145 rows
```

ahmvdev commented on sunpy/net/fido_factory.py:
```diff
@@ -186,7 +186,8 @@ def __str__(self):
         for block in self:
             ret += f"{len(block)} Results from the {block.client.__class__.__name__}:\n"
-            if block.client.info_url is not None:
+            if block.client.__class__.__name__ == 'VSOClient' and block.client.info_url is not None:
```
I may be misunderstanding the best way to implement this. Looking at the code:

1. I see that the output formatting is handled by __str__ in fido_factory.py:
```
  def __str__(self):
        nprov = len(self)
        if nprov == 1:
            ret = f'Results from {len(self)} Provider:\n\n'
        else:
            ret = f'Results from {len(self)} Providers:\n\n'
        for block in self:
            ret += f"{len(block)} Results from the {block.client.__class__.__name__}:\n"
            if block.client.info_url is not None:
                ret += f'Source: {block.client.info_url}\n'
            size = block.total_size()
```


2. My current understanding is that modifying VSOClient's info_url property wouldn't work well because fido_factory.py prepends "Source:" to the entire info_url output.

Could you help me understand:
1. Is this the correct interpretation of how the output formatting works?
2. What would be the best approach to add the health status link before the "Source:" line?

Thanks

ahmvdev replied:
Ah okay! Thanks for clarifying. I will look into this before pushing any other commit wildly.

nabobalis commented on changelog/7884.feature.rst:
```diff
@@ -0,0 +1 @@
+Add a link to the VSO Health Report in the Fido Results when using the VSOClient. (`#7853 <https://github.com/sunpy/sunpy/issues/7853>`__)
```
```suggestion
Add a link to the VSO Health Report in the Fido Results when using the VSOClient.
```

nabobalis commented:
Thanks for the PR @ahmvdev

ahmvdev commented:
> Thanks for the PR @ahmvdev

I realize i took on more than i could handle with this issue, possibly because i'm still a beginner. I really appreciate the constructive feedback from you and the supportive atmosphere of the community. I'm excited to continue learning and contributing, and i’m looking forward to getting more familiar with the codebase.
